### ISSUE TITLE
Server-Side Apply: ExtractPod should specify `UID`

### ISSUE THREAD

**Issue by gnuletik** [kubernetes/client-go#1346]
## TLDR

When using server side apply with ExtractPod, a controller may recreate a resource that was deleted by attempting to remove a finalizer.

## Context

I have a controller that watches pods to remove finalizer using Server-Side Apply (to avoid race condition between controllers [as suggested here](https://github.com/kubernetes/kubernetes/issues/111643)).

It removes the finalizer in the following way:
```go
podApplyConfig, err := coreac.ExtractPod(pod, fm)
if err != nil {
        return fmt.Errorf("coreac.ExtractPod: %w", err)
}

// updating something
podApplyConfig.Finalizers = nil

_, err = w.client.CoreV1().Pods(ns).Apply(ctx, podApplyConfig, meta.ApplyOptions{FieldManager: smtg})
if err != nil {
        return fmt.Errorf("client.Apply: %w", err)
}
```

The `SharedInformer` sometimes sends the event multiple times, which leads to pod being recreated after deletion, given this scenario:

- api-server: sets deletionTimestamp
- my-controller: receives Update event: it removes the finalizer with SSA
- api-server: deleted the pod
- my-controller: receives a **duplicated** Update event: it removes the finalizer a second time with SSA
- api-server: creates a pod instead of updating it

## Fix

Kubernetes specs specify that:

> Create is blocked on apply if uid is provided [link](https://github.com/kubernetes/kubernetes/blob/1b8c8f1daf4b1ed6d17ee1d2f40d62c8ecec0e15/test/integration/apiserver/apply/apply_test.go#L123-L154)

[See KEPS](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/555-server-side-apply/README.md#testing-plan)

We can manually call `podApply := podApply.WithUID(pod.UID)` to workaround this.

However, I think that it would be better that `ExtractPod` (and similar method for other resources) sets the UID before returning it.

https://github.com/kubernetes/client-go/blob/aa7909e7d7c0661792ba21b9e882f3cd6ad0ce53/applyconfigurations/core/v1/pod.go#L72-L84

WDYT?

I can make a PR if that sounds good to you.
However, this part of the codebase seems to be generated, so I'd need a hint on where to edit that.

Thanks!

**Comment by guettli** [comment]
It is a bit sad, that there was no single reply to that issue. I don't face the issue at the moment, but I have concerns to run into that sooner or later.

@gnuletik did you find an answer somewhere else?

**Comment by gnuletik** [comment]
@guettli When using `ExtractPod` (or similar functions), I manually set the `UID` with the following function:

```go
podApply := podApply.WithUID(pod.UID)
```

I also stopped using `ExtractPod` functions (and similar) when I can, and instead I "rebuild" the apply specifications.

**Comment by guettli** [comment]
@gnuletik just out of curiosity, how do you "rebuild" the apply spec?

I do it like that. But I am unsure if this is good, today is the first time that I use SSA from Go.

```go
	secret := &corev1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name:      containerRegistrySecretName,
			Namespace: mgtSystemNamespace,
		},
		TypeMeta: metav1.TypeMeta{
			APIVersion: "v1",
			Kind:       "Secret",
		},
		Type: corev1.SecretTypeDockerConfigJson,
		Data: map[string][]byte{
			".dockerconfigjson": alm.secret.Data[".dockerconfigjson"],
		},
	}
	data, err := json.Marshal(secret)
	if err != nil {
		return fmt.Errorf("failed to marshal secret: %w", err)
	}

	_, err = clientSet.CoreV1().Secrets(secret.Namespace).Patch(
		ctx,
		secret.Name,
		types.ApplyPatchType,
		data,
		metav1.PatchOptions{
			FieldManager: fieldManagerName,
		},
	)
	if err != nil {
		return fmt.Errorf("failed to apply secret: %w", err)
	}
```
